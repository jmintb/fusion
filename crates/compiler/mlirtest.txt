module attributes {llvm.data_layout = ""} {
  llvm.mlir.global internal @annon("hello") {addr_space = 0 : i32}
  llvm.func @main() attributes {llvm.emit_c_interface} {
    %0 = llvm.mlir.addressof @annon : !llvm.ptr
    %1 = llvm.call @printf(%0) : (!llvm.ptr) -> i32
    llvm.return
  }
  llvm.func @_mlir_ciface_main() attributes {llvm.emit_c_interface} {
    llvm.call @main() : () -> ()
    llvm.return
  }
  llvm.func @printf(!llvm.ptr, ...) -> i32 attributes {llvm.emit_c_interface, sym_visibility = "private"}
}