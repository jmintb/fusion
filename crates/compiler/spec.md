# Fusion Language Specification

version: 0.1
