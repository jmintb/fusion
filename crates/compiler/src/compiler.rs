use crate::analysis::pipeline::transform_ir;
use anyhow::Result;
use melior::ExecutionEngine;
use tracing::debug;

use crate::{
    ast::{identifiers::ScopeID, parser::parse, scopes::build_program_scopes},
    backend::mlir::codegen::{generate_mlir, MlirGenerationConfig},
    cli::load_program,
    cli::load_program_without_std_lib,
    ir::{IrGenerator, IrProgram},
    types::resolve_types,
};

use crate::analysis::type_evaluation::evaluate_types;

pub fn produce_ir(src: &str) -> Result<IrProgram> {
    let input = load_program(Some(src.to_string()))?;
    let (ast, mut node_db, _messages) = parse(&input)?;
    let program_scopes = build_program_scopes(&ast, &mut node_db);
    let (expression_types, type_db) = resolve_types(&ast, &node_db, &program_scopes, ScopeID(0));
    let ir_generator = IrGenerator::new(ast, node_db, expression_types, type_db);
    Ok(ir_generator.convert_to_ssa())
}

pub fn produce_ir_without_std(src: &str) -> Result<IrProgram> {
    let input = load_program_without_std_lib(Some(src.to_string()))?;
    let (ast, mut node_db, _messages) = parse(&input)?;
    let program_scopes = build_program_scopes(&ast, &mut node_db);
    let (expression_types, type_db) = resolve_types(&ast, &node_db, &program_scopes, ScopeID(0));
    debug!("type db: {:#?}", expression_types);
    let ir_generator = IrGenerator::new(ast, node_db, expression_types, type_db);
    Ok(ir_generator.convert_to_ssa())
}

pub fn compile_with_analysis(input: &str) -> Result<ExecutionEngine> {
    let ir = produce_ir(input)?;
    let transformed_ir = transform_ir(ir)?;
    let types = evaluate_types(&transformed_ir)?;

    // TODO: missing statement id to scope mapping

    let mlir_generation_config = MlirGenerationConfig {
        program: transformed_ir,
        verify_mlir: true,
        program_types: types,
    };

    let engine = generate_mlir(mlir_generation_config)?;
    // Overall goal introduce scopes and an ID based approach NOT a whole type system.
    // TODO: be able to output IR -> stay focus on e2e pipeline
    // 2. IR -> MLIR
    Ok(engine)
}

pub fn compile(input: &str) -> Result<ExecutionEngine> {
    let ir = produce_ir(input)?;
    let types = evaluate_types(&ir)?;

    // TODO: missing statement id to scope mapping

    let mlir_generation_config = MlirGenerationConfig {
        program: ir,
        verify_mlir: true,
        program_types: types,
    };

    let engine = generate_mlir(mlir_generation_config)?;
    // Overall goal introduce scopes and an ID based approach NOT a whole type system.
    // TODO: be able to output IR -> stay focus on e2e pipeline
    // 2. IR -> MLIR
    Ok(engine)
}

pub fn jit(input: &str) -> Result<()> {
    let ir = produce_ir(input)?;
    let types = evaluate_types(&ir)?;

    // TODO: missing statement id to scope mapping

    let mlir_generation_config = MlirGenerationConfig {
        program: ir,
        verify_mlir: true,
        program_types: types,
    };

    let engine = generate_mlir(mlir_generation_config)?;
    // Overall goal introduce scopes and an ID based approach NOT a whole type system.
    // TODO: be able to output IR -> stay focus on e2e pipeline
    // 2. IR -> MLIR
    Ok(unsafe { engine.invoke_packed("main", &mut []) }?)
}
