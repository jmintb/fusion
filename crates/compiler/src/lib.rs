mod analysis;
pub mod ast;
pub mod backend;
pub mod cli;
pub mod codegen;
pub mod compiler;
pub mod control_flow_graph;
pub mod identifiers;
mod ir;
pub mod parser;
pub mod typed_ast;
pub mod types;
