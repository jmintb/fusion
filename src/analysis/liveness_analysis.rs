use std::collections::{BTreeMap, BTreeSet};

use anyhow::{Result};

use tracing::debug;

use crate::{
    control_flow_graph::ControlFlowGraph,
    ir::{BlockId, FunctionId, Instruction, IrProgram, SSAID}, ast::identifiers::FunctionDeclarationID,
};

use super::ir_transformer::IrInterpreter;

#[derive(Clone, Debug, Copy)]
pub struct AbstractAddress {
    pub block_id: BlockId,
    pub inststruction: usize,
}

#[derive(Clone, Debug)]
pub struct AbstractAddressRange {
    pub start_address: AbstractAddress,
    pub end_addresses: Vec<AbstractAddress>,
}

impl AbstractAddressRange {
    fn insert_end(&mut self, address: AbstractAddress) {
        self.end_addresses.push(address);
    }

    fn new(start_address: AbstractAddress) -> Self {
        Self {
            start_address,
            end_addresses: vec![start_address],
        }
    }
}

#[derive(Clone, Debug, Default)]
pub struct VariableLiveness {
    pub variables: BTreeMap<SSAID, AbstractAddressRange>,
    pub variable_moved: BTreeMap<SSAID, BTreeSet<BlockId>>,
    pub loans: BTreeMap<SSAID, Vec<SSAID>>,
}

impl VariableLiveness {
    fn new() -> Self {
        Self {
            variables: BTreeMap::new(),
            loans: BTreeMap::new(),
            variable_moved: BTreeMap::new(),
        }
    }

    fn insert_variable_start(
        &mut self,
        variable_id: SSAID,
        address: AbstractAddress,
    ) -> Result<()> {
        self.variables
            .entry(variable_id)
            .and_modify(|entry| entry.start_address = address)
            .or_insert(AbstractAddressRange::new(address));

        Ok(())
    }

    fn insert_move(&mut self, variable_id: SSAID, block_id: BlockId) {
        self.variable_moved
            .entry(variable_id)
            .and_modify(|drop_ids| {
                drop_ids.insert(block_id);
            })
            .or_insert(BTreeSet::from_iter(vec![block_id].into_iter()));
    }

    fn insert_variable_end(
        &mut self,
        id: SSAID,
        address: AbstractAddress,
        moved: bool,
        control_flow_graph: &ControlFlowGraph<BlockId>,
    ) -> Result<()> {
        let _variables = self.variables.clone();
        let Some(entry) = self.variables.get_mut(&id) else {
            self.variables
                .insert(id, AbstractAddressRange::new(address));
            if moved {
                self.insert_move(id, address.block_id);
            }
            return Ok(());
        };

        if let Some(_end_in_current_block) = entry
            .end_addresses
            .iter()
            .position(|end_address| end_address.block_id == address.block_id)
        {
            return Ok(());
        } else {
            let entry_clone = entry.clone();
            if entry_clone
                .end_addresses
                .iter()
                .enumerate()
                .any(|(_, end_address)| {
                    control_flow_graph
                        .cycle_aware_successors(&address.block_id)
                        .unwrap()
                        .into_iter()
                        .flatten()
                        .collect::<Vec<BlockId>>()
                        .contains(&end_address.block_id)
                })
            {
                return Ok(());
            }
            entry.end_addresses.push(address);
            if moved {
                self.insert_move(id, address.block_id);
            }
        }

        Ok(())
    }

    // TODO: next use reverse traverself to calcuate liveness. If two blocks diverge keep end addresses.

    pub fn variabled_moved(&self, id: &SSAID, block_id: BlockId) -> bool {
        self.variable_moved
            .get(id)
            .map(|block_ids| block_ids.contains(&block_id))
            .unwrap_or(false)
    }
}

pub fn calculate_livenss(ir_program: &IrProgram) -> Result<BTreeMap<FunctionDeclarationID, VariableLiveness>> {
    debug!("calculating livenss for program: {ir_program}");

    let mut liveness_for_fn = BTreeMap::new();
    for (function_id, control_flow_graph) in ir_program.control_flow_graphs.clone() {
        debug!(
            "control flow graph: {}: {}\n predecessors: {:#?}",
            function_id.0,
            control_flow_graph,
            control_flow_graph
                .cycle_aware_successors(&control_flow_graph.entry_point)
                .unwrap()
        );
        let interpreter = IrInterpreter::<VariableLiveness>::new_reversed(
            ir_program
                .control_flow_graphs
                .get(&function_id)
                .cloned()
                .unwrap(),
            ir_program.clone(),
        );

        let livenss = interpreter.transform(
            &mut |instruction_counter, ctx, block_id, variable_livness| {
                debug!(
                    "liveness for {} at instruction {}",
                    block_id, instruction_counter
                );
                let block = ctx.scope.blocks.get(block_id).unwrap();
                let Some(instruction) = block.instructions.get(instruction_counter) else {
                    return Ok(block.instructions.len());
                };

                match instruction {
                    Instruction::Assign(to, from) => variable_livness.insert_variable_start(
                        *to,
                        AbstractAddress {
                            block_id: *block_id,
                            inststruction: instruction_counter,
                        },
                    )?,
                    Instruction::BorrowEnd(id) | Instruction::MutBorrowEnd(id) => {
                        variable_livness.insert_variable_end(
                            *id,
                            AbstractAddress {
                                block_id: *block_id,
                                inststruction: instruction_counter,
                            },
                            false,
                            &ctx.scope.control_flow_graph,
                        )?;
                    }
                    Instruction::Move(id) | Instruction::Drop(id) => {
                        variable_livness.insert_variable_end(
                            *id,
                            AbstractAddress {
                                block_id: *block_id,
                                inststruction: instruction_counter,
                            },
                            true,
                            &ctx.scope.control_flow_graph,
                        )?;
                    }
                    Instruction::Call(_, _, to) => {
                    variable_livness.insert_variable_start(
                        *to,
                        AbstractAddress {
                            block_id: *block_id,
                            inststruction: instruction_counter,
                        },
                    )?
                    }
                    _ => (),
                }

                Ok(if instruction_counter == 0 {
                    block.instructions.len()
                } else {
                    instruction_counter - 1
                })
            },
        )?;

        liveness_for_fn.insert(function_id.clone(), livenss);
    }

    Ok(liveness_for_fn)
}

#[cfg(test)]
mod test {
    use crate::cli::load_program;

    use super::*;
    use crate::ir::IrGenerator;
    use anyhow::Result;
    use rstest::rstest;
    use std::path::PathBuf;

    #[rstest]
    fn test_liveness(#[files("./ir_test_programs/test_*.ts")] path: PathBuf) -> Result<()> {
        use crate::{ast::{parser::parse, identifiers::ScopeID, scopes::build_program_scopes}, types::resolve_types, compiler::produce_ir};

        let ir_program = produce_ir(path.to_str().unwrap())?;
        let analysis_result = calculate_livenss(&ir_program);

        insta::assert_snapshot!(
            format!(
                "test_liveness_{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{:#?}", analysis_result)
        );


        insta::assert_snapshot!(
            format!(
                "test_liveness_{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{:#?}", analysis_result)
        );

        Ok(())
    }
}
