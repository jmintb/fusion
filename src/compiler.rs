use anyhow::Result;

use crate::{
    ast::{identifiers::ScopeID, parser::parse, scopes::build_program_scopes},
    ir::{IrGenerator, IrProgram},
    types::resolve_types, cli::load_program,
};

pub fn produce_ir(src: &str) -> Result<IrProgram> {
    let input = load_program(Some(src.to_string()))?;
    let (ast, mut node_db, messages) = parse(&input)?;
    let program_scopes = build_program_scopes(&ast, &mut node_db);
    let (expression_types, type_db) = resolve_types(&ast, &node_db, &program_scopes, ScopeID(0));
    let ir_generator = IrGenerator::new(ast, node_db, program_scopes, expression_types, type_db);
    Ok(ir_generator.convert_to_ssa())
}

pub fn jit(input: &str) -> Result<()> {
    let ir = produce_ir(input);

    // TODO: missing statement id to scope mapping

    let engine = todo!();
    // Overall goal introduce scopes and an ID based approach NOT a whole type system.
    // TODO: be able to output IR -> stay focus on e2e pipeline
    // 2. IR -> MLIR
    // let engine = generate_mlir(typed_program, false)?;
    // Ok(unsafe { engine.invoke_packed("main", &mut []) }?)
}
