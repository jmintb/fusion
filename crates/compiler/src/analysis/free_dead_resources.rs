use std::collections::BTreeMap;

use tracing::debug;

use crate::{
    analysis::liveness_analysis::AbstractAddress,
    ast::identifiers::FunctionDeclarationID,
    ir::{BlockId, Instruction, IrProgram},
};

use super::liveness_analysis::VariableLiveness;

pub fn insert_free(
    variable_liveness: BTreeMap<FunctionDeclarationID, VariableLiveness>,
    mut ir_program: IrProgram,
) -> IrProgram {
    let mut block_offsets = BTreeMap::<BlockId, usize>::new();
    debug!(
        "variable liveness: {:#?} \n program {}",
        variable_liveness, ir_program
    );
    for variable_liveness in variable_liveness.into_values() {
        for (variable_id, address) in variable_liveness.clone().variables {
            debug!("variable: {}, address: {:#?}, ", variable_id.0, address);
            assert!(!address.end_addresses.is_empty());
            for end_address in address.end_addresses {
                if variable_liveness.variabled_moved(&variable_id, end_address.block_id) {
                    continue;
                }
                let block_offset = block_offsets.entry(end_address.block_id).or_insert(0);
                let target_block_id = end_address.block_id;
                let Some(cfg) = ir_program
                    .control_flow_graphs
                    .clone()
                    .into_values()
                    .find(|cfg| cfg.contains(&target_block_id))
                else {
                    debug!("failed to find cfg containing {}", target_block_id.0);

                    continue;
                };
                let mut offset = 1;
                let address = if cfg.is_in_cycle(&target_block_id) {
                    // debug!("{} is in cycle", variable_id.0);
                    let target_block_id = cfg
                        .find_cycle_successor(&target_block_id)
                        .unwrap_or_else(|| {
                            panic!(
                                "expected to find a successor to the cycle in {} \n {} \n {}",
                                target_block_id,
                                ir_program,
                                ir_program.entry_point_cfg()
                            )
                        });
                    offset = 0;
                    AbstractAddress {
                        block_id: target_block_id,
                        inststruction: 0,
                    }
                } else {
                    debug!("{} is not in cycle", variable_id.0);
                    end_address
                };

                debug!(
                    "inserting drop for variable {} in block {} at line {} with offset {}",
                    variable_id.0, address.block_id, address.inststruction, offset
                );

                ir_program
                    .blocks
                    .get_mut(&address.block_id)
                    .unwrap()
                    .instructions
                    .insert(
                        address.inststruction + offset + *block_offset,
                        Instruction::Drop(variable_id),
                    );

                block_offsets
                    .entry(address.block_id)
                    .and_modify(|offset| *offset += 1)
                    .or_insert(0);
            }
        }
    }

    ir_program
}

#[cfg(test)]
mod test {

    use anyhow::{bail, Result};
    use rstest::rstest;
    use std::path::PathBuf;

    #[rstest]
    #[test_log::test]
    fn test_ir_output_with_drops(
        #[files("./ir_test_programs/test_*.ts")] path: PathBuf,
    ) -> Result<()> {
        use crate::compiler::produce_ir;

        let ir_program = produce_ir(path.to_str().unwrap())?;
        let liveness = crate::analysis::liveness_analysis::calculate_livenss(&ir_program)?;
        let ir_program = crate::analysis::free_dead_resources::insert_free(liveness, ir_program);

        insta::assert_snapshot!(
            format!(
                "test_well_formed_ir_after_frees_{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{}", ir_program)
        );
        Ok(())
    }

    #[rstest]
    #[test_log::test]
    fn test_all_variables_are_dropped(
        #[files("./ir_test_programs/test_*.ts")] path: PathBuf,
    ) -> Result<()> {
        use crate::compiler::produce_ir;

        let ir_program = produce_ir(path.to_str().unwrap())?;
        let liveness = crate::analysis::liveness_analysis::calculate_livenss(&ir_program)?;
        let ir_program = crate::analysis::free_dead_resources::insert_free(liveness, ir_program);
        let liveness = crate::analysis::liveness_analysis::calculate_livenss(&ir_program)?;

        for variable_liveness in liveness.into_values() {
            for (variable_id, end_addresses) in variable_liveness.variables.clone() {
                for end_address in end_addresses.end_addresses {
                    if !variable_liveness.variabled_moved(&variable_id, end_address.block_id) {
                        bail!("variable {} was not moved in {}", variable_id.0, ir_program);
                    }
                }
            }
        }

        Ok(())
    }
}
