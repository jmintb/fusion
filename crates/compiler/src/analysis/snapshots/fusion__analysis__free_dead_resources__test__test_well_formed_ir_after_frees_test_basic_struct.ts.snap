---
source: crates/compiler/src/analysis/free_dead_resources.rs
expression: "format!(\"{}\", ir_program)"
---
IR:
fn 17:
block_6
0: 7_val = fnarg
1: 8_len = fnarg
2: 9_anonymous := Int(1)
3: borrow(9_anonymous)
4: 10_anonymous := "w"
5: borrow(10_anonymous)
6: receiver_Ssaid(11) = @0( let 9_anonymous, let 10_anonymous,)
7: borrow_end(9_anonymous)
8: borrow_end(10_anonymous)
9: 12_stdoutptr = 11_fdopen_result
10: drop(9_anonymous)
11: move(11_fdopen_result)
12: drop(10_anonymous)
13: borrow(7_val)
14: borrow(8_len)
15: 13_anonymous := Int(1)
16: borrow(13_anonymous)
17: borrow(12_stdoutptr)
18: receiver_Ssaid(14) = @2( let 7_val, let 8_len, let 13_anonymous, let 12_stdoutptr,)
19: borrow_end(7_val)
20: drop(7_val)
21: borrow_end(8_len)
22: drop(8_len)
23: drop(14_fwrite_result)
24: borrow_end(13_anonymous)
25: borrow_end(12_stdoutptr)
26: drop(13_anonymous)
27: drop(12_stdoutptr)
28: return

fn 42:
block_15
0: 16_anonymous := "232"
1: 17_bar = 16_anonymous
2: drop(17_bar)
3: move(16_anonymous)
4: 18_anonymous := Int(60)
5: 19_anonymous := Int(40)
6: struct_init_result_Ssaid(20) = StructInit<10>[ 18_anonymous, 19_anonymous,]
7: 21_aFoo = 20_struct_init_result
8: drop(21_aFoo)
9: move(20_struct_init_result)
10: struct_field_read_resul_Ssaid(22) = Ssaid(21)_FieldRead.2
11: struct_field_read_resul_Ssaid(23) = Ssaid(21)_FieldRead.3
12: 24_@addition_result = 22_struct_field_result + 23_struct_field_result
13: move(22_struct_field_result)
14: move(23_struct_field_result)
15: 25_sum = 24_@addition_result
16: move(24_@addition_result)
17: 26_anonymous := Int(59)
18: 27_@greater_than_result = 25_sum > 26_anonymous
19: move(25_sum)
20: move(26_anonymous)
21: if 27_@greater_than_result then 28_block

block_28
0: 29_anonymous := "struct fields sum is 6"
1: borrow(29_anonymous)
2: 30_anonymous := Int(22)
3: borrow(30_anonymous)
4: @17( let 29_anonymous, let 30_anonymous,)
5: borrow_end(29_anonymous)
6: drop(29_anonymous)
7: borrow_end(30_anonymous)
8: drop(30_anonymous)
9: 31_anonymous := "\n"
10: borrow(31_anonymous)
11: 32_anonymous := Int(2)
12: borrow(32_anonymous)
13: @17( let 31_anonymous, let 32_anonymous,)
14: borrow_end(31_anonymous)
15: drop(31_anonymous)
16: borrow_end(32_anonymous)
17: drop(32_anonymous)

block_33
0: return
